File: framework/languages/java/samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/TicTacToeServerApplicationGroup.java
package systems.zlink.samples.tictactoe.server;

import org.springframework.context.ConfigurableApplicationContext;
import systems.zlink.samples.tictactoe.server.api.ApiServerApplication;
import systems.zlink.samples.tictactoe.server.configuration.SampleSettings;
import systems.zlink.samples.tictactoe.server.play.PlayServerApplication;

public final class TicTacToeServerApplicationGroup {
    private TicTacToeServerApplicationGroup() {
    }

    public static AutoCloseable run(SampleSettings settings) {
        ConfigurableApplicationContext play = null;
        ConfigurableApplicationContext api = null;
        try {
            play = PlayServerApplication.run(settings);
            api = ApiServerApplication.run(settings);
            return new ServerHost(play, api);
        } catch (RuntimeException error) {
            close(api);
            close(play);
            throw error;
        }
    }

    private record ServerHost(
        ConfigurableApplicationContext play,
        ConfigurableApplicationContext api) implements AutoCloseable {
        @Override
        public void close() {
            TicTacToeServerApplicationGroup.close(api);
            TicTacToeServerApplicationGroup.close(play);
        }
    }

    private static void close(ConfigurableApplicationContext context) {
        if (context != null) {
            context.close();
        }
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/TicTacToeServerApplicationGroup.kt
package systems.zlink.samples.kotlin.tictactoe.server

import org.springframework.context.ConfigurableApplicationContext
import systems.zlink.samples.kotlin.tictactoe.server.api.ApiServerApplication
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleSettings
import systems.zlink.samples.kotlin.tictactoe.server.play.PlayServerApplication

object TicTacToeServerApplicationGroup {
    fun run(settings: SampleSettings): AutoCloseable {
        var play: ConfigurableApplicationContext? = null
        var api: ConfigurableApplicationContext? = null
        try {
            play = PlayServerApplication.run(settings)
            api = ApiServerApplication.run(settings)
            return ServerHost(play, api)
        } catch (error: RuntimeException) {
            api?.close()
            play?.close()
            throw error
        }
    }

    private class ServerHost(
        private val play: ConfigurableApplicationContext?,
        private val api: ConfigurableApplicationContext?,
    ) : AutoCloseable {
        override fun close() {
            api?.close()
            play?.close()
        }
    }
}
